### PR TITLE
Revert "tests: Support calling add_nodes more than once"

### PR DESCRIPTION
**Pull request by maflcko**
Writing tests should be straightforward and with little side-effects as possible.

I don't see how this is needed and can not be achieved with `self.num_nodes` (and `self.extra_args` et al.)

### CODE REVIEW

**Comment by jnewbery** [comment]
utACK faa831102a0ce11b673b72081eed159ad0c17159

reasons: https://github.com/bitcoin/bitcoin/pull/14805#issuecomment-447021398

sorry @stevenroose !

**Comment by laanwj** [comment]
I'm not sure this is worth reverting, but okay

**Comment by jnewbery** [comment]
Can you add a comment to `add_nodes()` explaining why it should only be called once.

**Comment by MarcoFalke** [comment]
Is there a reason other than it being the design we want it to? I added a comment to mention that it should be called once after the parameters for all nodes have been set up.

**Comment by jnewbery** [comment]
> Is there a reason other than it being the design we want it to?

That's really it. Some of the test framework infrastructure relies on the `num_nodes` not changing. We *could* change that, but I think declaring the number of nodes at the start of the test is a good thing to do anyway.

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -292,8 +295,19 @@ def add_nodes(self, num_nodes, extra_args=None, *, rpchost=None, binary=None):
         assert_equal(len(extra_args), num_nodes)
         assert_equal(len(binary), num_nodes)
         for i in range(num_nodes):
-            numnode = len(self.nodes)
-            self.nodes.append(TestNode(numnode, get_datadir_path(self.options.tmpdir, numnode), rpchost=rpchost, timewait=self.rpc_timewait, bitcoind=binary[i], bitcoin_cli=self.options.bitcoincli, mocktime=self.mocktime, coverage_dir=self.options.coveragedir, extra_conf=extra_confs[i], extra_args=extra_args[i], use_cli=self.options.usecli))
+            self.nodes.append(TestNode(
```
Comment (laanwj) [inline test/functional/test_framework/test_framework.py]: *much* better readable

**Comment by jnewbery** [comment]
> So the correct was is to call add_nodes once and build a list of the binaries, chain names, etc beforehand?

Yes. If necessary, the node's arguments and config file can be updated later with `TestNode.args` and `append_config()`.

**Pull request merged**